algebras/algebra/src/main/scala/endpoints4s/algebra/AuthenticatedEndpoints.scala

```diff
@@ -0,0 +1,176 @@
+      )
+    )
+
+  /** Add Basic HTTP authentication to an endpoint.
+    *
+    * @param realm the `"realm"` attribute of the `"WWW-Authenticate"` response header. Defaults to `"Realm"`.
+    */
+  def endpointWithBasicAuth[A, B](
+      endpoint: Endpoint[A, B],
+      realm: Option[String] = None
+  )(implicit tupler: Tupler[A, UsernamePassword]): Endpoint[tupler.Out, Option[B]] =
+    endpoint
+      .mapRequest(_.withHeaders(basicAuthHeader))
+      .mapResponse(_.orElse(unauthorizedResponse(realm)).xmap(_.left.toOption)(_.toLeft(())))
+
+  /** Add Bearer token authentication to an endpoint.
+    *
+    * @param bearerFormat the format of the bearer token. Used in OpenApi.
+    */
+  def endpointWithBearerAuth[A, B](
+      endpoint: Endpoint[A, B],
+      bearerFormat: Option[String] = None
+  )(implicit tupler: Tupler[A, BearerToken]): Endpoint[tupler.Out, Option[B]] =
+    endpoint
+      .mapRequest(_.withHeaders(bearerAuthHeader))
+      .mapResponse(_.orElse(unauthorizedResponse(None)).xmap(_.left.toOption)(_.toLeft(())))
+
+  implicit class AuthenticationEndpointOps[A, B](endpoint: Endpoint[A, B]) {
+
+    /** Add Basic HTTP authentication to this endpoint.
+      *
+      * @param realm the `"realm"` attribute of the `"WWW-Authenticate"` response header. Defaults to `"Realm"`.
+      */
+    def withBasicAuth(
+        realm: Option[String] = None
+    )(implicit tupler: Tupler[A, UsernamePassword]): Endpoint[tupler.Out, Option[B]] =
+      endpointWithBasicAuth(endpoint, realm)
+
+    /** Add Bearer token authentication to this endpoint.
+      *
+      * @param bearerFormat the format of the bearer token. Used in OpenApi.
+      */
+    def withBearerAuth(
+        bearerFormat: Option[String] = None
+    )(implicit tupler: Tupler[A, BearerToken]): Endpoint[tupler.Out, Option[B]] =
+      endpointWithBearerAuth(endpoint)
+  }
+}
+object AuthenticatedEndpoints {
+  case class Credentials(username: String, password: String)
+  case class Bearer(token: String)
+}
+
+/** Interpreter for [[AuthenticatedEndpoints]] on the client-side.
+  *
+  * @group interpreters
+  */
+trait AuthenticatedEndpointsClient extends AuthenticatedEndpoints {
+
+  type UsernamePassword = AuthenticatedEndpoints.Credentials
+  lazy val basicAuthHeader: RequestHeaders[UsernamePassword] =
+    requestHeader("Authorization")
+      .xmapWithCodec(basicAuthCodec)
+      .xmap(AuthenticatedEndpoints.Credentials.tupled)(credentials =>
+        credentials.username -> credentials.password
+      )
+
+  type BearerToken = AuthenticatedEndpoints.Bearer
+  lazy val bearerAuthHeader: RequestHeaders[BearerToken] =
+    requestHeader("Authorization")
+      .xmapWithCodec(bearerAuthCodec)
+      .xmap(AuthenticatedEndpoints.Bearer)(_.token)
+}
+
+/** Interpreter for [[AuthenticatedEndpoints]] on the server-side.
+  *
+  * @group interpreters
+  */
+trait AuthenticatedEndpointsServer extends AuthenticatedEndpoints {
```

Comment:
One concern I have is that we don’t really encapsulate the fact that if the `Authorization` header is missing, then the server returns an `Unauthorized` response, otherwise it forwards to the underlying endpoint logic.

IIUC, in the current implementation we always call the underlying endpoint logic. I believe it would be interesting to be able to intercept requests.

Maybe we need something different from `mapRequest` and `mapResponse`.

Reply:
Indeed, a client could send a request without the `Authorization` header and a giant payload and the server would tries to decode the payload and only after reject the request because the header is missing. That's kind of a security vulnerability.

I guess, the implementations of `addRequestHeaders` and `addRequestQueryString` could be made such that they first extract the added headers/queryString and call the extraction of the rest of the request only if the former succeeded.

Reply:
I think it would be useful to control whether to call or not the underlying endpoint logic. A bit like what they do in Play, with the `ActionFunction` hierarchy.